rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,659 @@
+enforcement for, the versioning of teleport installations.
+
+4. Support a wide range of use-cases by providing a flexible and extensible set of tools
+with support for things like user-provided upgrade scripts and custom target selection.
+
+5. Offer a simple "batteries included" automatic upgrade option that requires minimal
+configuration and "just works".
+
+
+### Design Philosophy
+
+- Level-triggered loops
+- Idempotence
+- 
+
+### Abstract Model Overview
+
+This document proposes an modular system capable of supporting a wide range
+of upgrade strategies, with the intention being that the default or "batteries
+included" upgrade strategy will be implemented as a set of interchangeable
+components, each of which could be swapped out for some alternative implementation.
+
+The proposed system consists of at least the following components:
+
+- `version-directive`: A static resource that describes the desired state of
+versioning across the cluster. The directive includes matchers which allow
+the auth server to match individual teleport instances with both the appropriate
+installation target, and the appropriate installation method. This resource may
+be periodically generated by teleport or by some custom external program. It may
+also be manually created by an administrator.
+
+- `version-controller`: A pluggable component responsible for generating the
+`version-directive` based on some dynamic state (e.g. a server which publishes
+package versions and hashes). A builtin `version-controller` would simply be a
+configuration resource from the user's perspective. A custom/external version
+controller would be any program with the permissions necessary to update the
+`version-directive` resource.
+
+- *version reconciliation loop*: A control loop that runs in the auth server which
+comparse the desired state as specified by the `version-directive` with the current
+state of teleport installations across the cluster. When mismatches are discovered,
+the appropriate `install-controller`s are run.
+
+- `install-controller`: A component capable of attempting to affect installation
+of a specific target on one or more teleport hosts. The auth server needs to know
+enough to at least start the process, but the core logic of a given `install-controller`
+may be an external command (e.g. the proposed `local-script` controller would cause
+each teleport instance in need of upgrade to run a user-supplied script locally and
+then restart). From the perspective of a user, an `install-controller` is a teleport
+configuration object (though that configuration object may only be a thin hook into
+the "real" `install-controller`). Whether or not the teleport instance being upgraded
+needs to understand the `install-controller` will depend on the specific controller.
+
+There is room in the above model for additional granularity (e.g. schedulers), but
+the listed components are a decent starting point.
```

Comment:
I agree with the model of converging to a state by a pluggable controller. It is very close (modeled after?) Kubernetes controllers.